### PR TITLE
Remove sleep statements from `login-page.js`.

### PR DESCRIPTION
**Pull request by juliaamosova**
In this PR, I removed 5 sleep statements and refactored the code to substitute for them. 

Working only some of the time, **sleep statements are not deterministic**. As a result, we end up with fragile tests that break under certain conditions.

**Sleep 1:**

Refactored `waitTillPresentAndDisplayed()` and removed `sleep()` stament that followed it. 

Instead, I used `driver.wait()` method to wait for up to 20 seconds for the element to be found in the DOM. If the element is added to the DOM in one second, the `driver.wait()` method will complete in one second. Unlike `driver.sleep()`, it will not wait the full amount of time specified. 

Another advantage of using `driver.wait()` method is that in case of an error, the error message will be shown specifying what went wrong. 

**Sleep 2:**

Removed sleep statements after `await driverHelper.setWhenSettable( driver, userNameSelector, username );`. It is redundant because the statement below completes user's input. 

**Sleep 3:**

Removed and replaced with `driver.wait()` method. The explanation is the same as for Sleep 1. 

**Sleep 4:**

Same explanation as for Sleep 2.

**Sleep 5:**

Removed sleep statement as it is redundant because the statement below `waitTillNotPresent()` performs the waiting needed. 

-----------

I also doubled the wait time in `await driverHelper.waitTillNotPresent( driver, userNameSelector, this.explicitWaitMS * 2 );` as the test was failing locally due to timeout. Sometimes, it takes up to 30 seconds for the test to be completed on my machine:

Image Link: https://cloudup.com/cCAuLGMqJ8d

Short video showing the test passing locally:

http://g.recordit.co/YBqADKXgJE.gif

I tested on tablet and mobile locally as well - passed.

### CODE REVIEW

File: lib/pages/login-page.js [hunk 1]
```diff
@@ -21,29 +22,32 @@ export default class LoginPage extends AsyncBaseContainer {
 		const userNameSelector = By.css( '#usernameOrEmail' );
 		const passwordSelector = By.css( '#password' );
 
-		await driverHelper.waitTillPresentAndDisplayed( driver, userNameSelector );
-		await this.driver.sleep( 1000 );
+		await this.driver.wait(
```
Comment (bsessions85) [inline lib/pages/login-page.js]: If this really does work better than `waitTillPresentAndDisplayed`, then it would probably be better to update that rather than just fixing in this one spot

File: lib/pages/login-page.js [hunk 2]
```diff
@@ -21,29 +22,32 @@ export default class LoginPage extends AsyncBaseContainer {
...
+			this.explicitWaitMS,
+			'Could not locate the username or email field.'
+		);
+
 		await driverHelper.setWhenSettable( driver, userNameSelector, username );
-		await this.driver.sleep( 1000 );
 		await driver.findElement( userNameSelector ).sendKeys( Key.ENTER );
 
-		await this.driver.sleep( 1000 );
+		await this.driver.wait(
```
Comment (bsessions85) [inline lib/pages/login-page.js]: Would `waitTillPresentAndDisplayed` work here instead? Reuse is always preferred

**Comment by JavonDavis** [comment]
Additionally I see you're working on the Login specs, there was some instability with these noted here -> https://github.com/Automattic/wp-e2e-tests/issues/995. As a result, these tests aren't ran in the full suite in the CI anymore to avoid errors due to the session still being active. The `Needs Secure Auth e2e Testing` label triggers these tests to run in the Ci. That being said when you're testing locally you'll need to space out your executions to avoid this or disable the 2fa tests (marked with @secure-auth) if you're testing more frequently.

**Comment by juliaamosova** [comment]
>Additionally I see you're working on the Login specs, there was some instability with these noted here -> #995. As a result, these tests aren't ran in the full suite in the CI anymore to avoid errors due to the session still being active. The Needs Secure Auth e2e Testing label triggers these tests to run in the Ci. That being said when you're testing locally you'll need to space out your executions to avoid this or disable the 2fa tests (marked with @secure-auth) if you're testing more frequently.

Thanks for these details and insight @JavonDavis, very helpful and good to know. Thanks for adding the `Needs Secure Auth e2e Testing` label as well to trigger the test!

File: lib/pages/login-page.js (hunk 1 above)
Comment (juliaamosova) [reply lib/pages/login-page.js]: Thanks for double checking on this @bsessions85. I've done additional research and confirm that it is better to use `waitTillPresentAndDisplayed` as it provides more coverage.

- `driver.wait(until.elementLocated)` - checks that an element is _present_ within the HTML DOM of a page. This does not necessarily mean that the element is _visible_. So there's no guarantee that it's _interactable_.

- `waitTillPresentAndDisplayed` function - includes `driver.findElement()` and `element.isDisplayed()` - in other words, it checks that an element is _present_ within the HTML DOM of a page and is _visible_. This does not necessarily mean that the element is _interactable_ or _clickable_ but at least it is already _visible_. 

>Part of the problem with removing sleeps is that they are often times put in on a day when things are running slower or the tests start failing suddenly. A slight delay is added and things start passing again. That being said, it is important to think through what might be happening that caused the sleep to be added in the first place.

`waitTillPresentAndDisplayed` is followed by `setWhenSettable` which finds the element, clears the text from the textarea and enters the value. 

I checked why the sleep statement was added here https://github.com/Automattic/wp-e2e-tests/pull/1528 - tests run too fast and a delay is added. While there is no explanation of why the test is failing exactly, the reason could be that there is no check if the cursor had been placed in the input field or if the input field has a focus. If the test runs too fast, the input field may be disaplyed but not in focus yet while the driver is trying to clear it already. 

I am working on testing my theory above. Will submit another PR as soon as I am done.

File: lib/pages/login-page.js (hunk 2 above)
Comment (juliaamosova) [reply lib/pages/login-page.js]: Yes, as per my research in https://github.com/Automattic/wp-e2e-tests/pull/1698#discussion_r240554182, it is better to use `waitTillPresentAndDisplayed`. I will address it in the next PR. Thanks @bsessions85!

File: lib/driver-helper.js [hunk 3]
```diff
@@ -45,6 +45,55 @@ export function clickWhenClickable( driver, selector, waitOverride ) {
 	);
 }
 
+export function clickWhenFocused( driver, selector, pollingOverride, waitOverride ) {
```
Comment (juliaamosova) [inline lib/driver-helper.js]: As per my comment here https://github.com/Automattic/wp-e2e-tests/pull/1698#discussion_r240554182 the reason why the tests that run too fast are failing could be that there is no check if the cursor had been placed in the input field or if the input field has a focus. The input field may be disaplyed but not in focus yet while the driver is trying to clear it already.

To address this issue I wrote a new `driverHelper` method `clickWhenFocused`. It is based on the following steps:

- Find element;
- Make it active (in focus) by switching to it;
- Make sure that this element equals the element on which the next action must be performed;
- Click on that element.

The focus is implemented by using `switchTo().activeElement()`. 

`activeElement()` switches to the element that currently has focus within the document currently "switched to", or the body element if this cannot be detected. See reference [here](https://seleniumhq.github.io/selenium/docs/api/java/org/openqa/selenium/WebDriver.TargetLocator.html#activeElement--). 

Once the element is in focus we need to make sure that it matches the element that needs to be clicked on. To do that, we can utilize `getId()` which is a server assigned `Id` to the element of the DOM of the particualr session (not the HTML `Id`) as per reference [here](https://github.com/SeleniumHQ/selenium/blob/master/javascript/node/selenium-webdriver/lib/webdriver.js#L1829).  

@bsessions85 let me know your thoughts. Also cc @Stojdza as you wanted to work on re-writing `clickWhenClickable` with a similar approach.

File: lib/pages/login-page.js [hunk 4]
```diff
@@ -22,28 +22,23 @@ export default class LoginPage extends AsyncBaseContainer {
 		const passwordSelector = By.css( '#password' );
 
 		await driverHelper.waitTillPresentAndDisplayed( driver, userNameSelector );
-		await this.driver.sleep( 1000 );
+		await driverHelper.clickWhenFocused( driver, userNameSelector );
```
Comment (juliaamosova) [inline lib/pages/login-page.js]: Implementing a new `driverHelper` method `clickWhenFocused` to address the sleep statement that was previously here.

File: lib/pages/login-page.js [hunk 5]
```diff
@@ -22,28 +22,23 @@ export default class LoginPage extends AsyncBaseContainer {
...
 		await driverHelper.setWhenSettable( driver, userNameSelector, username );
-		await this.driver.sleep( 1000 );
 		await driver.findElement( userNameSelector ).sendKeys( Key.ENTER );
 
-		await this.driver.sleep( 1000 );
-
 		if ( emailSSO === false ) {
+			await driverHelper.waitTillPresentAndDisplayed( driver, passwordSelector );
+			await driverHelper.clickWhenFocused( driver, passwordSelector );
```
Comment (juliaamosova) [inline lib/pages/login-page.js]: Implementing `waitTillPresentAndDisplayed` and a new `driverHelper` method `clickWhenFocused` to address the sleep statement that was previously here.

File: lib/driver-helper.js (hunk 3 above)
Comment (Stojdza) [reply lib/driver-helper.js]: @juliaamosova This method is also applicable to buttons and other elements? If yes, did you consider to use it as `clickWhenClickable` method (rewrite existing one)? Those two look similar, and it's easier just to change the existing one and use it through all tests.

Comment (juliaamosova) [reply lib/driver-helper.js]: @Stojdza I did consider re-writing `clickWhenClickable` but didn't want to use it before the review of the new method I wrote. I agree that keeping `clickWhenClickable` will be better. The method I wrote should be applicable to all types of elements. Let me test it locally and I will get back with a new PR.

Comment (juliaamosova) [reply lib/driver-helper.js]: @Stojdza I refactored `clickWhenClickable` with the changes I introduced in my `clickWhenFocused`method. While it is working locally, I see that some tests are failing here mainly due to timeouts. I am looking to see what could be going on.

File: lib/driver-helper.js [hunk 6]
```diff
@@ -45,6 +45,38 @@ export function clickWhenClickable( driver, selector, waitOverride ) {
 	);
 }
 
+export function waitTillFocused( driver, selector, pollingOverride, waitOverride ) {
```
Comment (juliaamosova) [inline lib/driver-helper.js]: @bsessions85 and @Stojdza so far, I addressed 2 out of 5 sleep statements in this PR. I used `waitTillFocused` method that I wrote. I remove the `click` from it (originally had it) since it is not needed. 
I placed the remaining 3 sleep statements that I originally removed back for now.
I think it is best if I address the remaining 3 in this file in a separate PR since the existing one already has a long history and it is getting hard to follow.

File: lib/driver-helper.js (hunk 3 above)
Comment (juliaamosova) [reply lib/driver-helper.js]: @Stojdza after additional testing, I can't confirm that I can re-write `clickWhenClickable` with changes I introduced in `waitTillFocused` method. It seems to work in many cases but not always. I removed the `click` event from the  `waitTillFocused` so that it could be used to make sure the element is in focus only. It works great to address the sleep statements in `login-page.js`.

File: lib/pages/login-page.js [hunk 7]
```diff
@@ -22,28 +22,27 @@ export default class LoginPage extends AsyncBaseContainer {
...
+		await driverHelper.waitTillFocused( driver, userNameSelector );
 		await driverHelper.setWhenSettable( driver, userNameSelector, username );
 		await this.driver.sleep( 1000 );
 		await driver.findElement( userNameSelector ).sendKeys( Key.ENTER );
 
-		await this.driver.sleep( 1000 );
-
 		if ( emailSSO === false ) {
+			await driverHelper.waitTillPresentAndDisplayed( driver, passwordSelector );
+			await driverHelper.waitTillFocused( driver, passwordSelector );
```
Comment (Stojdza) [inline lib/pages/login-page.js]: Is it necessary to have both methods, `waitTillPresentAndDisplayed` and `waitTillFocused`? It seems that we are waiting for the same element twice.

Comment (juliaamosova) [reply lib/pages/login-page.js]: @Stojdza I tested without `waitTillPresentAndDisplayed` here and on line 24 as well (https://github.com/Automattic/wp-e2e-tests/pull/1698/files#diff-d2cabd92e3f34ad1459c4dd1862e098fR24). It works without it. But I decided to leave it because `waitTillFocused` doesn't check if the element is displayed and `waitTillPresentAndDisplayed` has that check. I think to be extra sure that the element is displayed, it will be good to leave it. What do you think?

>I wanted to try what would happen if I put waitTillFocused inside clickWhenClickable and setWhenSettable but that did not turn out well :D

I did the same but just with `clickWhenClickable` and it was an epic fail.

Comment (juliaamosova) [reply lib/pages/login-page.js]: @Stojdza thanks! Does it mean it is good to merge?

Comment (Stojdza) [reply lib/pages/login-page.js]: Yes :)

Comment (juliaamosova) [reply lib/pages/login-page.js]: Cool, thanks!
